source/parser/x86_64/type_checker.hpp

```diff
@@ -27,12 +29,46 @@ namespace smeagle::x86_64 {
   inline bool is_ref(st::dataClass dc) { return dc == st::dataReference; }
 
   // Return true if dataClass is reference or pointer
-  inline bool is_indirect(st::dataClass dc) {
-    return dc == st::dataPointer || dc == st::dataReference;
-  }
+  inline bool is_indirect(st::dataClass dc) { return is_pointer(dc) || is_ref(dc); }
 
   // Return true if dataClass is a primitive type
-  inline bool is_primitive(st::dataClass dc) { return dc == st::dataEnum || dc == st::dataScalar; }
+  inline bool is_primitive(st::dataClass dc) { return dc == st::dataScalar; }
   inline bool is_typedef(st::dataClass dc) { return dc == st::dataTypedef; }
+  inline bool is_enum(st::dataClass dc) { return dc == st::dataEnum; }
+
+  // Dereference a pointer or reference
+  inline st::Type *deref(st::Type *t) {
+    if (is_pointer(t->getDataClass())) {
+      return deref(t->getPointerType()->getConstituentType());
+    }
+    if (is_ref(t->getDataClass())) {
+      return deref(t->getRefType()->getConstituentType());
+    }
+    return t;
+  }
+  // Unwrap and remove typedef
+  inline st::Type *remove_typedef(st::Type *t) {
+    if (is_typedef(t->getDataClass())) {
+      t = t->getTypedefType()->getConstituentType();
+    }
+    return t;
+  };
+
+  namespace detail {
```

Comment:
Why namespace detail, and not something like namespace type_checker?

Reply:
The overload of `dedecorate` that takes two parameters should only be called from `dedecorate` that takes one parameter. Because we're in a header file, I "hid" this function in the `detail` namespace. It's a fairly common idiom- the Boost libraries overrun with the `detail` namespace.